Python/Product/Analysis/AnalysisUnit.cs

```diff
@@ -421,8 +423,51 @@ public ClassAnalysisUnit(ClassDefinition node, InterpreterScope declScope, Analy
                 }
             }
 
+
             ddg.SetCurrentUnit(this);
             ddg.WalkBody(Ast.Body, classInfo.AnalysisUnit);
+
+            ddg.SetCurrentUnit(_outerUnit);
+            var v = _outerUnit.Scope.AddLocatedVariable(Ast.Name, Ast.NameExpression, this);
+            v.AddTypes(this, ProcessClassDecorators(ddg, classInfo));
+        }
+
+        private IAnalysisSet ProcessClassDecorators(DDG ddg, ClassInfo classInfo) {
+            var types = classInfo.SelfSet;
+            if (Ast.Decorators != null) {
+                Expression expr = Ast.NameExpression;
+
+                foreach (var d in Ast.Decorators.DecoratorsInternal) {
+                    if (d != null) {
+                        var decorator = ddg._eval.Evaluate(d);
+
+                        Expression nextExpr;
+                        if (!_decoratorCalls.TryGetValue(d, out nextExpr)) {
+                            nextExpr = _decoratorCalls[d] = new CallExpression(d, new[] { new Arg(expr) });
+                            nextExpr.SetLoc(d.IndexSpan);
+                        }
+                        expr = nextExpr;
+                        var decorated = AnalysisSet.Empty;
+                        bool anyResults = false;
+                        foreach (var ns in decorator) {
+                            var fd = ns as FunctionInfo;
+                            if (fd != null && Scope.EnumerateTowardsGlobal.Any(s => s.AnalysisValue == fd)) {
+                                continue;
+                            }
+                            decorated = decorated.Union(ns.Call(expr, this, new[] { types }, ExpressionEvaluator.EmptyNames));
```

Comment:
The use of `types` here is what propagates the result of the previous iteration? There doesn't seem to be any test for that  though?